src/ray/util/event.h

```diff
@@ -0,0 +1,177 @@
+
+ protected:
+  virtual std::string EventToString(rpc::Event &event);
+ 
+ protected:
+  const std::string separator_ = "|||";
+  std::string file_name_;
+  std::string log_dir_; 
+};
+
+class EventManager {
+ public:
+  static EventManager &Instance() {
+    static EventManager INSTANCE;
+    return INSTANCE;
+  }
+
+  bool IsEmpty() { return reporter_map_.empty(); }
+
+  void Publish(rpc::Event &event) {
+    for (const auto &element : reporter_map_) {
+      (element.second)->Report(event);
+    }
+  }
+
+  void AddReporter(std::shared_ptr<LogBasedEventReporter> reporter) {
+    reporter_map_.emplace(reporter->GetReporterKey(), reporter);
+  }
+
+  void ClearReporters() { reporter_map_.clear(); }
+
+ private:
+  EventManager() {}
+
+  EventManager(const EventManager &manager) = delete;
+
+  const EventManager &operator=(const EventManager &manager) = delete;
+
+ protected:
+  std::unordered_map<std::string, std::shared_ptr<LogBasedEventReporter>> reporter_map_;
+};
+
+class RayEventContext {
+ public:
+  static RayEventContext &Instance() {
+    if (instance_ == nullptr) {
+      instance_ = std::unique_ptr<RayEventContext>(new RayEventContext());
+    }
+    return *instance_;
+  }
+
+  // SetEventContext should only be called once in a process, unless ResetEventContext is
+  // called.
+  void SetEventContext(rpc::Event_SourceType source_type, std::string node_id = "",
+                       std::string job_id = "", std::string task_id = "");
+
+  void ResetEventContext();
+
+  inline void SetJobID(std::string job_id) { job_id_ = job_id; }
+
+  inline void SetNodeID(std::string node_id) { node_id_ = node_id; }
+
+  inline void SetTaskID(std::string task_id) { task_id_ = task_id; }
+
+  inline std::string GetJobID() { return job_id_; }
+
+  inline std::string GetNodeID() { return node_id_; }
+
+  inline std::string GetTaskID() { return task_id_; }
+
+  inline rpc::Event_SourceType GetSourceType() { return source_type_; }
+
+  inline std::string GetSourceHostname() { return source_hostname_; }
+
+  inline int32_t GetSourcePid() { return source_pid_; }
+
+ private:
+  RayEventContext() {}
+
+  RayEventContext(const RayEventContext &event_context) = delete;
+
+  const RayEventContext &operator=(const RayEventContext &manaevent_contextger) = delete;
+
+ private:
+  std::string job_id_ = "";
+  std::string node_id_ = "";
+  std::string task_id_ = "";
+  rpc::Event_SourceType source_type_ = rpc::Event_SourceType::Event_SourceType_COMMON;
+  std::string source_hostname_ = boost::asio::ip::host_name();
+  int32_t source_pid_ = getpid();
+
+  static thread_local std::unique_ptr<RayEventContext> instance_;
+};
+
+class RayEvent {
+ public:
+  RayEvent(rpc::Event_Severity severity, std::string label)
+      : severity_(severity), label_(label) {}
+
+  template <typename T>
+  RayEvent &operator<<(const T &t) {
+    osstream_ << t;
+    return *this;
+  }
+
+  static void ReportEvent(std::string severity, std::string label, std::string message) {
```

Comment:
cc @raulchen @edoakes 

Is it the best practice to just send messages for events? In this case, I really cannot find differences between logs and events except that the cardinality is lower. I had some impression that we should somehow pre-define events  with some description, so that we can prevent using tons of events in the repo? Idk what's the best practice in the industry, so let me know what you think about this.